src/System.Drawing.Common/src/System/Drawing/Image.cs

```diff
@@ -314,24 +301,57 @@ internal void SetNativeImage(IntPtr handle)
             nativeImage = handle;
         }
 
-        internal static void EnsureSave(Image image, string filename, Stream dataStream)
+        // Multi-frame support
+
+        /// <summary>
+        /// Gets an array of GUIDs that represent the dimensions of frames within this <see cref='Image'/>.
+        /// </summary>
+        [Browsable(false)]
+        public unsafe Guid[] FrameDimensionsList
+        {
+            get
+            {
+                Gdip.CheckStatus(Gdip.GdipImageGetFrameDimensionsCount(new HandleRef(this, nativeImage), out int count));
+
+                Debug.Assert(count >= 0, "FrameDimensionsList returns bad count");
+                if (count <= 0)
+                    return Array.Empty<Guid>();
+
+                Guid[] guids = new Guid[count];
+                fixed (Guid* g = guids)
+                {
+                    Gdip.CheckStatus(Gdip.GdipImageGetFrameDimensionsList(new HandleRef(this, nativeImage), g, count));
+                }
+
+                return guids;
+            }
+        }
+
+        internal unsafe static void EnsureSave(Image image, string filename, Stream dataStream)
         {
             if (image.RawFormat.Equals(ImageFormat.Gif))
             {
                 bool animatedGif = false;
 
-                Guid[] dimensions = image.FrameDimensionsList;
-                foreach (Guid guid in dimensions)
+                Gdip.CheckStatus(Gdip.GdipImageGetFrameDimensionsCount(new HandleRef(image, image.nativeImage), out int dimensions));
+                if (dimensions <= 0)
                 {
-                    FrameDimension dimension = new FrameDimension(guid);
-                    if (dimension.Equals(FrameDimension.Time))
+                    return;
+                }
+
+                Guid* guids = stackalloc Guid[dimensions];
```

Comment:
do we know that the dimensions are small enough to always stackalloc?

Reply:
I'm not sure, I'll update to fall back to the array if crazy.